<a href="https://colab.research.google.com/github/Marisol787/Marisol787-MarisolGutierrezDTSC3020/blob/main/Assignment_6_WebScraping__(1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment 6 (4 points) — Web Scraping

In this assignment you will complete **two questions**. The **deadline is posted on Canvas**.


## Assignment Guide (Read Me First)

- This notebook provides an **Install Required Libraries** cell and a **Common Imports & Polite Headers** cell. Run them first.
- Each question includes a **skeleton**. The skeleton is **not** a solution; it is a lightweight scaffold you may reuse.
- Under each skeleton you will find a **“Write your answer here”** code cell. Implement your scraping, cleaning, and saving logic there.
- When your code is complete, run the **Runner** cell to print a Top‑15 preview and save the CSV.
- Expected outputs:
  - **Q1:** `data_q1.csv` + Top‑15 sorted by the specified numeric column.
  - **Q2:** `data_q2.csv` + Top‑15 sorted by `points`.


In [2]:
# Install Required Libraries
!pip -q install requests beautifulsoup4 lxml pandas
print("Dependencies installed.")

Dependencies installed.


### 2) Common Imports & Polite Headers

In [3]:
# Common Imports & Polite Headers
import re, sys, pandas as pd, requests
from bs4 import BeautifulSoup
HEADERS = {"User-Agent": (
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/122.0 Safari/537.36")}
def fetch_html(url: str, timeout: int = 20) -> str:
    r = requests.get(url, headers=HEADERS, timeout=timeout)
    r.raise_for_status()
    return r.text
def flatten_headers(df: pd.DataFrame) -> pd.DataFrame:
    if isinstance(df.columns, pd.MultiIndex):
        df.columns = [" ".join([str(x) for x in tup if str(x)!="nan"]).strip()
                      for tup in df.columns.values]
    else:
        df.columns = [str(c).strip() for c in df.columns]
    return df
print("Common helpers loaded.")


Common helpers loaded.


## Question 1 — IBAN Country Codes (table)
**URL:** https://www.iban.com/country-codes  
**Extract at least:** `Country`, `Alpha-2`, `Alpha-3`, `Numeric` (≥4 cols; you may add more)  
**Clean:** trim spaces; `Alpha-2/Alpha-3` → **UPPERCASE**; `Numeric` → **int** (nullable OK)  
**Output:** write **`data_q1.csv`** and **print a Top-15** sorted by `Numeric` (desc, no charts)  
**Deliverables:** notebook + `data_q1.csv` + short `README.md` (URL, steps, 1 limitation)

**Tip:** You can use `pandas.read_html(html)` to read tables and then pick one with ≥3 columns.


In [41]:
# --- Q1 Skeleton (fill the TODOs) ---
def q1_read_table(html: str) -> pd.DataFrame:
    """Return the first table with >= 3 columns from the HTML.
    TODO: implement with pd.read_html(html), pick a reasonable table, then flatten headers.
    """
    raise NotImplementedError("TODO: implement q1_read_table")

def q1_clean(df: pd.DataFrame) -> pd.DataFrame:
    """Clean columns: strip, UPPER Alpha-2/Alpha-3, cast Numeric to int (nullable), drop invalids.
    TODO: implement cleaning steps.
    """
    raise NotImplementedError("TODO: implement q1_clean")

def q1_sort_top(df: pd.DataFrame, top: int = 15) -> pd.DataFrame:
    """Sort descending by Numeric and return Top-N.
    TODO: implement.
    """
    raise NotImplementedError("TODO: implement q1_sort_top")


In [42]:
# Q1 — Write your answer here

import re, sys, pandas as pd, requests
from bs4 import BeautifulSoup
HEADERS = {"User-Agent": (
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/122.0 Safari/537.36")}

def fetch_html(url: str, timeout: int = 20) -> str:
    r = requests.get(url, headers=HEADERS, timeout=timeout)
    r.raise_for_status()
    return r.text
def flatten_headers(df: pd.DataFrame) -> pd.DataFrame:
    if isinstance(df.columns, pd.MultiIndex):
        df.columns = [" ".join([str(x) for x in tup if str(x)!="nan"]).strip()
                      for tup in df.columns.values]
    else:
        df.columns = [str(c).strip() for c in df.columns]
    return df

Q1LINK ="https://www.iban.com/country-codes"

def q1_read_table(html: str) -> pd.DataFrame:
    """Return first table with >=3 columns using pandas.read_html."""
    tables = pd.read_html(html)
    for t in tables:
        if t.shape[1] >= 3:
            df = flatten_headers(t)
            return df
    raise ValueError("no valid table found with >=3 columns.")


def q1_clean(df: pd.DataFrame) -> pd.DataFrame:
    df.columns = [c.strip().title() for c in df.columns]
    df = df.rename(columns={
        "Alpha-2 code":"Alpha-2",
        "Alpha-3 code":"Alpha-3",
        "Numeric code":"Numeric"
    })
    for col in ["Country", "Alpha-2", "Alpha-3"]:
        if col in df.columns:
            df[col] = df[col].astype(str).str.strip().str.upper()
    if "Numeric" in df.columns:
        df["Numeric"] = pd.to_numeric(df["Numeric"], errors="coerce").astype("Int64")
    df = df.dropna(subset=["Country"])
    return df

def q1_sort_top(df: pd.DataFrame, top: int = 15) -> pd.DataFrame:

    return df.sort_values(by="Numeric", ascending=False, na_position="last").head(top)
html1 = fetch_html(Q1LINK)
df1 = q1_read_table(html1)
df1 = q1_clean(df1)
TOP15 = q1_sort_top(df1)
print("TOP 15 IBAN Country Codes!!")
print(TOP15.to_string(index=False))
df1.to_csv("data_q1.csv", index=False)
print("\nSaved as data_q1.csv\n")




TOP 15 IBAN Country Codes!!
                                                   Country Alpha-2 Code Alpha-3 Code  Numeric
                                                    ZAMBIA           ZM          ZMB      894
                                                     YEMEN           YE          YEM      887
                                                     SAMOA           WS          WSM      882
                                         WALLIS AND FUTUNA           WF          WLF      876
                        VENEZUELA (BOLIVARIAN REPUBLIC OF)           VE          VEN      862
                                                UZBEKISTAN           UZ          UZB      860
                                                   URUGUAY           UY          URY      858
                                              BURKINA FASO           BF          BFA      854
                                     VIRGIN ISLANDS (U.S.)           VI          VIR      850
                            UNIT

/tmp/ipython-input-3888476293.py:25: FutureWarning: Passing literal html to 'read_html' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  tables = pd.read_html(html)


## Question 2 — Hacker News (front page)
**URL:** https://news.ycombinator.com/  
**Extract at least:** `rank`, `title`, `link`, `points`, `comments` (user optional)  
**Clean:** cast `points`/`comments`/`rank` → **int** (non-digits → 0), fill missing text fields  
**Output:** write **`data_q2.csv`** and **print a Top-15** sorted by `points` (desc, no charts)  
**Tip:** Each story is a `.athing` row; details (points/comments/user) are in the next `<tr>` with `.subtext`.


In [18]:
# --- Q2 Skeleton (fill the TODOs) ---
def q2_parse_items(html: str) -> pd.DataFrame:
    """Parse front page items into DataFrame columns:
       rank, title, link, points, comments, user (optional).
    TODO: implement with BeautifulSoup on '.athing' and its sibling '.subtext'.
    """
    raise NotImplementedError("TODO: implement q2_parse_items")

def q2_clean(df: pd.DataFrame) -> pd.DataFrame:
    """Clean numeric fields and fill missing values.
    TODO: cast points/comments/rank to int (non-digits -> 0). Fill text fields.
    """
    raise NotImplementedError("TODO: implement q2_clean")

def q2_sort_top(df: pd.DataFrame, top: int = 15) -> pd.DataFrame:
    """Sort by points desc and return Top-N. TODO: implement."""
    raise NotImplementedError("TODO: implement q2_sort_top")


In [34]:
# Q2 — Write your answer here
URL_Q2 = "https://news.ycombinator.com/"

def q2_parse_items(html: str) -> pd.DataFrame:
    """Extract rank, title, link, points, comments, user from Hacker News."""
    soup = BeautifulSoup(html, "lxml")
    news = soup.select("tr.athing")
    rows = []
    for s in news:
        rank = s.select_one(".rank")
        title = s.select_one(".titleline a")
        subtext = s.find_next_sibling("tr").select_one(".subtext")
        points = subtext.select_one(".score") if subtext else None
        user = subtext.select_one(".hnuser") if subtext else None
        comments = None
        if subtext:
            cm = subtext.find_all("a")
            if cm:
                for a in cm[::-1]:
                    if "comment" in a.text:
                        comments = a
                        break
        rows.append({
            "rank": rank.text.replace(".", "").strip() if rank else "",
            "title": title.text.strip() if title else "",
            "link": title.get("href") if title else "",
            "points": points.text.strip() if points else "",
            "comments": comments.text.strip() if comments else "",
            "user": user.text.strip() if user else "",
        })
    return pd.DataFrame(rows)


def q2_clean(df: pd.DataFrame) -> pd.DataFrame:
    for col in ["rank", "points", "comments"]:
        df[col] = (
            df[col].astype(str)
            .str.extract(r"(\d+)")[0]
            .fillna(0)
            .astype(int)
        )
    df = df.fillna("")
    return df

def q2_sort_top(df: pd.DataFrame, top: int = 15) -> pd.DataFrame:
    """Return Top sorted by points desc."""
    return df.sort_values(by="points", ascending=False).head(top)

html2 = fetch_html(URL_Q2)
df2 = q2_parse_items(html2)
df2 = q2_clean(df2)
top15_q2 = q2_sort_top(df2)

print("TOP 15 Hacker News ")
print(top15_q2.to_string(index=False))
df2.to_csv("data_q2.csv", index=False)
print("\nSaved as data_q2.csv\n")



TOP 15 Hacker News 
 rank                                                                       title                                                                                                                                                                link  points  comments         user
   12 YouTube Removes Windows 11 Bypass Tutorials, Claims 'Risk of Physical Harm'                                                                                               https://news.itsfoss.com/youtube-removes-windows-11-bypass-tutorials/     398       148  WaitWaitWha
   10                                                     Why I love OCaml (2023)                                                                                                                     https://mccd.space/posts/ocaml-the-worlds-best/     296       201        art-w
   20              VLC's Jean-Baptiste Kempf Receives the European SFS Award 2025                                                                    

# Task
Implement the `q2_parse_items`, `q2_clean`, and `q2_sort_top` functions to extract `rank`, `title`, `link`, `points`, and `comments` from the Hacker News front page ("https://news.ycombinator.com/"), clean the data (casting numeric fields to int, handling non-digits as 0, and filling missing text fields), sort by `points` in descending order, print the top 15 results, and save the entire cleaned DataFrame to "data_q2.csv".

## Implement q2_parse_items

### Subtask:
Implement the `q2_parse_items` function to parse the Hacker News HTML, extracting the rank, title, link, points, and comments for each story into a pandas DataFrame. Use BeautifulSoup to navigate the HTML structure, specifically targeting `.athing` rows and their subsequent `.subtext` rows.


**Reasoning**:
Implementing the `q2_parse_items` function as per the instructions, which involves parsing the HTML content using BeautifulSoup to extract specific data points (rank, title, link, points, comments, and optional user) and organizing them into a pandas DataFrame.



In [19]:
def q2_parse_items(html: str) -> pd.DataFrame:
    """Parse front page items into DataFrame columns:
       rank, title, link, points, comments, user (optional).
    """
    data = []
    soup = BeautifulSoup(html, "html.parser")

    athing_rows = soup.find_all("tr", class_="athing")

    for row in athing_rows:
        # Extract rank
        rank_tag = row.find("span", class_="rank")
        rank = int(rank_tag.get_text(strip=True).replace(".", "")) if rank_tag else None

        # Extract title and link
        title_line_tag = row.find("span", class_="titleline")
        title_tag = title_line_tag.find("a") if title_line_tag else None
        title = title_tag.get_text(strip=True) if title_tag else None
        link = title_tag.get("href") if title_tag else None

        # The subtext is typically in the next sibling tr
        subtext_tr = row.find_next_sibling("tr")
        subtext_td = subtext_tr.find("td", class_="subtext") if subtext_tr else None

        # Extract points
        points = 0
        score_tag = subtext_td.find("span", class_="score") if subtext_td else None
        if score_tag:
            points_text = score_tag.get_text(strip=True)
            match = re.search(r'(\d+)', points_text)
            if match:
                points = int(match.group(1))

        # Extract comments
        comments = 0
        comment_tag = None
        if subtext_td:
            # Look for an anchor tag that contains 'comments', 'comment', or 'discuss'
            # and also contains digits for the count.
            for a_tag in subtext_td.find_all("a"):
                if re.search(r'comments?|discuss', a_tag.get_text(strip=True), re.IGNORECASE):
                    comment_tag = a_tag
                    break

        if comment_tag:
            comment_text = comment_tag.get_text(strip=True)
            match = re.search(r'(\d+)', comment_text)
            if match:
                comments = int(match.group(1))

        # Extract user (optional)
        user = None
        user_tag = subtext_td.find("a", class_="hnuser") if subtext_td else None
        if user_tag:
            user = user_tag.get_text(strip=True)

        data.append({
            "rank": rank,
            "title": title,
            "link": link,
            "points": points,
            "comments": comments,
            "user": user
        })

    return pd.DataFrame(data)

**Reasoning**:
The next step is to implement the `q2_clean` function according to the instructions, which involves casting numeric fields (`points`, `comments`, `rank`) to integers (treating non-digits as 0) and filling any missing text fields.



In [20]:
def q2_clean(df: pd.DataFrame) -> pd.DataFrame:
    """Clean numeric fields and fill missing values.
    Cast points/comments/rank to int (non-digits -> 0). Fill text fields.
    """
    # Cast numeric fields to int, treating non-digits as 0
    for col in ["rank", "points", "comments"]:
        if col in df.columns:
            # Convert to numeric, coerce errors to NaN, then fill NaN with 0 and cast to Int64
            df[col] = pd.to_numeric(df[col], errors="coerce").fillna(0).astype('Int64')

    # Fill missing text fields with appropriate defaults
    for col in ["title", "link", "user"]:
        if col in df.columns:
            df[col] = df[col].fillna("N/A")

    return df

**Reasoning**:
Now that `q2_parse_items` and `q2_clean` are defined, the next step is to implement the `q2_sort_top` function as specified in the problem description, which involves sorting the DataFrame by `points` in descending order and returning the top N rows.



In [21]:
def q2_sort_top(df: pd.DataFrame, top: int = 15) -> pd.DataFrame:
    """Sort by points desc and return Top-N."""
    return df.sort_values(by="points", ascending=False).head(top)

**Reasoning**:
With all helper functions (`q2_parse_items`, `q2_clean`, and `q2_sort_top`) now defined, the next step is to execute the complete Question 2 pipeline: fetch the HTML from the specified URL, parse it using `q2_parse_items`, clean the resulting DataFrame using `q2_clean`, sort and select the top 15 entries using `q2_sort_top`, print the top 15, and save the full cleaned DataFrame to `data_q2.csv`.



In [22]:
Q2LINK = "https://news.ycombinator.com/"

html2 = fetch_html(Q2LINK)
df2 = q2_parse_items(html2)
df2 = q2_clean(df2)
TOP15_Q2 = q2_sort_top(df2)

print("Hacker News — Top 15 by Points")
print(TOP15_Q2[['rank', 'title', 'points', 'comments']].to_string(index=False))

df2.to_csv("data_q2.csv", index=False)
print("\nSaved as data_q2.csv")

Hacker News — Top 15 by Points
 rank                                                                         title  points  comments
   29 Denmark's government aims to ban access to social media for children under 15     416       300
   12   YouTube Removes Windows 11 Bypass Tutorials, Claims 'Risk of Physical Harm'     393       144
   10                                                       Why I love OCaml (2023)     296       201
   20                VLC's Jean-Baptiste Kempf Receives the European SFS Award 2025     240        39
   21                                                         James Watson has died     231       125
    7      Myna: Monospace typeface designed for symbol-heavy programming languages     202        77
    9                                                        Ruby Solved My Problem     171        68
    1                                                           Why is Zig so Cool?     144        44
   30                                              

## Final Task

### Subtask:
Summarize the completion of Question 2, including the successful extraction, cleaning, and saving of Hacker News data, along with the displayed Top-15.


## Summary:

### Data Analysis Key Findings
*   The `q2_parse_items` function successfully extracted `rank`, `title`, `link`, `points`, `comments`, and `user` for each story from the Hacker News front page.
*   The `q2_clean` function effectively cleaned the extracted data by casting numeric fields (`rank`, `points`, `comments`) to `Int64`, converting non-digit values to 0, and filling missing text fields (`title`, `link`, `user`) with "N/A".
*   The `q2_sort_top` function correctly sorted the cleaned data by `points` in descending order and identified the top 15 stories.
*   An example of a top story extracted is "Denmark's government aims to ban access to social media for children under 15" with 416 points and 300 comments.
*   The entire cleaned DataFrame was successfully saved to "data\_q2.csv".

### Insights or Next Steps
*   The implemented functions provide a reliable pipeline for periodically fetching, cleaning, and analyzing Hacker News data, which could be extended to track news trends or community engagement over time.
*   To enhance the analysis, one could incorporate sentiment analysis on story titles or comments to gauge public opinion, or track the performance of specific authors/users.
